src/console/man/om_evaluate.1
.\" Text automatically generated by txt2man
.TH om_evaluate  "09 December 2014" "" ""
.SH NAME
\fBom_evaluate \fP- return model values given a set of points and an environmental scenario using the openModeller framework
\fB
.SH SYNOPSIS
.nf
.fam C
     \fBom_evaluate\fP [-] \fIv\fP \fB--version\fP | \fIr\fP \fB--xml-req\fP \fIFILE\fP | \fIo\fP \fB--model\fP \fIFILE\fP \fIp\fP \fB--points\fP \fIFILE\fP  [ \fIs\fP \fB--result\fP \fIFILE\fP ] [ \fB--log-level\fP \fILEVEL\fP ] [ \fB--log-file\fP \fIFILE\fP ] [ \fB--prog-file\fP \fIFILE\fP ]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
\fBom_evaluate\fP is a command line tool to return raw model values given a set of points and an environmental scenario. There are two ways of providing input: one is to specify an XML file containing a request according to the ModelEvaluationParameters element definition in the openModeller XML Schema located in http://openmodeller.cria.org.br/xml/2.0/openModeller.xsd (the test_request.xml file in the openModeller examples directory can also be used with \fBom_evaluate\fP, although the root element name is related to the test operation) and the other is to specify a file with a serialized model according to the SerializedModel element definition in the openModeller XML Schema together with another file with the points to be tested (TAB-delimited, following the same pattern used to specify points in text files for om_console). When providing a serialized model and a file with points to be tested, these points must have the same spatial reference and label as the training points that can be found in the serialized model. In this case, the layers used during the test will also be the same ones used to create the model. Results will be either displayed on the screen or stored in another file if the corresponding parameter was specified. The result will be in XML, following the ModelEvaluationType definition in the openModeller XML Schema. Serialized models can be generated with om_console or om_model. When a point does not have the complete environmental values (one or more layer has no data), model evaluation returns \fB-1\fP.
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fIr\fP, \fB--xml-req\fP
File containing a test request in XML.
.TP
.B
-\fIo\fP, \fB--model\fP
File containing a serialized model in XML.
.TP
.B
-\fIp\fP, \fB--points\fP
File containing the points to be tested.
.TP
.B
-\fIs\fP, \fB--result\fP
File where the test result will be stored.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.TP
.B
\fB--log-file\fP
File to store log information.
.TP
.B
\fB--prog-file\fP
File to store progress (\fB-1\fP=queued, \fB-2\fP=aborted, \fB-3\fP=cancelled, [0,100]=progress).
.PP
\fB-c\fP, \fB--config-file\fP Configuration file for openModeller (available since version 1.4).
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 1.4.0 (?-?-?)

SVN revision: ?

	* New CONSENSUS algorithm that can be used to generate models with different algorithms by converting each individual model into a binary model given a threshold and then "merging" them into a final combined model according to a specified level of agreement.
	* Upgraded gSOAP version being used by OMWS to 2.8.15.
	* Included library API versioning, starting with 0 from this release.
	* Moved some of the compiler definitions from CMake to internal #defines, so that client programs can include openModeller headers without worrying about adding specific definitions if they want to compile directly without using CMake.
	* New command-line tool om_evaluate to return raw model values given a set of points and an environmental scenario.
	* Fixed segfault when needNormalization was called on a deserialized SVM algorithm.
	* Modified om_pseudo command-line tool to also accept XML input (in this case also producing XML output), optionally storing it into a file. When a model is specified as a parameter, presences are generated inside the model, while absences are generated outside.
	* New implementation of the SOAP client/server programs compatible with OMWS 2.0.
	* New driver to fetch remote rasters available through http, https or ftp. Layers can now be identified by URLs.
	* New class to read openModeller settings from a configuration file (key = value). Possible configurations include: CACHE_DIRECTORY, ALGS_DIRECTORY and DATA_DIRECTORY.
	* Changed initialization priority order for paths: 1) programatic setting, 2) configuration file, 3) environment variable, 4) default value through compiler constant.
	* Renamed compiler constant PKGDATAPATH and environment variable OM_DATA_PATH to OM_DATA_DIR.
	* Renamed compiler constant PLUGINPATH and environment variable OM_ALG_PATH to OM_ALGS_DIR.
	* New optional parameter --config-file in most command line tools to indicate a configuration file for openModeller: om_model, om_test, om_project, om_points, om_pseudo, om_evaluate, om_sampler and om_algorithm. om_console also accepts an optional third argument to specify the configuration file location.
	* New command-line tool om_layer to either check if a specified layer id can be read from openModeller or to recursively scan a specified directory looking for openModeller-compatible layers.
	* New configuration FETCH_WCS (true or false) allowing WCS rasters to be fetched and stored as local Erdas Imagine copies before being used.


--------------------------
Release 1.3.0 (2012-07-23)

SVN revision: 5592

	* New version of OM-MAXENT compatible with Maxent 3.3.3e.
	* Changed the deserialization code to accept XML with namespace prefixes (they are now ignored if present).
	* Fixed bug in ANN algorithm which was limiting the number of input points to 10k (seg faulting otherwise).
	* New documentation about the web service interface (OMWS), aka SOAP interface.
	* New documentation about oM Server (server-side implementation of OMWS).
	* Fixed bug in oM Server getProgress operation which was sometimes returning 100% when the job was still unfinished.
	* New oM Server configuration/script allowing automatic synchronization of layers.
	* Several improvements in the OMWS Perl command-line client.
	* Several adjustments in the SWIG Python interface, exposing more methods and objects.
	* Fixed memory leak in the Environmental Distance algorithm when the Mahalanobis metric was used after model deserialization.
	* New WCS proxy raster driver that allows WCS rasters to be identified through strings in the format: wcs>wcs_service_url>coverage_id


--------------------------
Release 1.2.0 (2011-07-21)

SVN revision: 5330

	* Changed the ENFA algorithm to force marginality to be the first factor.  Changed broken stick distribution to work on specialisation factors only.  Add new parameter to enable log display of variable loadings for retained factors.
	* om_console now accepts a second parameter (after the request file) to specify the log level.
	* Changed the ENFA algorithm allowing absence data to be treated as background points, and log transforming the geomean calculation to avoid the occasional float overflow with large numbers of presences. Fixed a bug in the RetainVariation parameter (range changed from 50-100 to 0.5-1.0). Also fixed a deserialization error.
	* Compilation now requires CMake 2.6.4.
	* New algorithm Random Forests.
	* Web service interface now accepts a new optional parameter specifying the prediction threshold to be used when calculating area statistics during projection.
	* Renamed envelope element in the getProjectionMetadata response of the web service API to distinguish it from the subelement.
	* Added control to a maximum number of concurrent jobs on the modelling server.
	* Environmental values can now be loaded from the Sampler XML when provided.
	* Web service interface now accepts optional parameters: IgnoreAbsences for the confusion matrix and UseAbsencesAsBackground for the ROC curve calculations.
	* Added support to gzip compression in the Web service interface.
	* Changed the Environmental Distance algorithm to return chi-square distribution probabilities when the metric is Mahalanobis and the maximum distance is 1.
	* Changed the Maxent algorithm allowing absence data to be treated as background points. Also fixed memory deallocation issues.


--------------------------
Release 1.1.0 (2010-02-05)

SVN revision: 5114

	* Bugfix: Interquartile and envelope size adjustments were using wrong reference values for salinity and temperature in the Aquamaps algorithm.
	* Bugfix: All GARP algorithms were crashing when the abundance value associated with one of the occurrence points had more than two digits in the integer part.
	* Increased the default number of neurons in the hidden layer of the ANN algorithm.
	* New argument "s" in om_niche allowing model outputs to be scaled.

src/console/om_evaluate.cpp
#include <openmodeller/om.hh>
#include <openmodeller/Exceptions.hh>
#include <openmodeller/os_specific.hh>

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"

#include <fstream>   // file I/O for XML
#include <sstream>   // ostringstream datatype
#include <stdio.h>   // file I/O for log
#include <time.h>    // used to limit the number of times that the progress is written to a file
#include <string>    // string library
#include <stdexcept> // try/catch

using namespace std;

int get_values(AlgorithmPtr alg, EnvironmentPtr env, OccurrencesPtr occs, Scalar * values, int cnt);

/// Main code
int main( int argc, char **argv ) {

  Options opts;
  int option;

  // command-line parameters (short name, long name, description, take args)
  opts.addOption( "v", "version"     , "Display version info"                           , false );
  opts.addOption( "r", "xml-req"     , "(option 1) Model evaluation request file in XML", true );
  opts.addOption( "o", "model"       , "(option 2) Serialized model file"               , true );
  opts.addOption( "p", "points"      , "(option 2) TAB-delimited file with points"      , true );
  opts.addOption( "s", "result"      , "File to store evaluation result in XML"         , true );
  opts.addOption( "" , "log-level"   , "Set the log level (debug, warn, info, error)"   , true );
  opts.addOption( "" , "log-file"    , "Log file"                                       , true );
  opts.addOption( "" , "prog-file"   , "File to store job progress"                     , true );
  opts.addOption( "c", "config-file" , "Configuration file for openModeller"            , true );

  std::string log_level("info");
  std::string request_file;
  std::string model_file;
  std::string points_file;
  std::string result_file;
  std::string log_file;
  std::string progress_file;
  std::string config_file;

  if ( ! opts.parse( argc, argv ) ) {

    opts.showHelp( argv[0] );
  }

  // Set up any related external resources
  setupExternalResources();

  OpenModeller om;

  while ( ( option = opts.cycle() ) >= 0 ) {

    switch ( option ) {

      case 0:
        printf( "om_evaluate %s\n", om.getVersion().c_str() );
        printf("This is free software; see the source for copying conditions. There is NO\n");
        printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
        exit(0);
        break;
      case 1:
        request_file = opts.getArgs( option );
        break;
      case 2:
        model_file = opts.getArgs( option );
        break;
      case 3:
        points_file = opts.getArgs( option );
        break;
      case 4:
        result_file = opts.getArgs( option );
        break;
      case 5:
        log_level = opts.getArgs( option );
        break;
      case 6:
        log_file = opts.getArgs( option );
        break;
      case 7:
        progress_file = opts.getArgs( option );
        break;
      case 8:
        config_file = opts.getArgs( option );
        break;
      default:
        break;
    }
  }

  // Initialize progress data if user wants to track progress
  progress_data prog_data;

  if ( ! progress_file.empty() ) { 

    prog_data.file_name = progress_file;

    time( &prog_data.timestamp );

    prog_data.progress = -1.0; // queued

    // Always create initial file with progress 0
    progressFileCallback( 0.0, &prog_data );
  }

  // Log stuff

  Log::Level level_code = getLogLevel( log_level );

  if ( ! log_file.empty() ) {

    Log::instance()->set( level_code, log_file, "" );
  }
  else {
 
    // Just set the level - things will go to stderr
    Log::instance()->setLevel( level_code );
  }

  // om configuration
  if ( ! config_file.empty() ) { 

    Settings::loadConfig( config_file );
  }

  // Check parameters

  if ( request_file.empty() ) {

    if ( model_file.empty() || points_file.empty() ) {

      printf( "Please specify either a test request file in XML or a serialized model and a TAB-delimited file with the points to be tested\n");

      // If user is tracking progress
      if ( ! progress_file.empty() ) { 

        // -2 means aborted
        progressFileCallback( -2.0, &prog_data );
      }

      exit(-1);
    }
  }
  else {

    if ( ! model_file.empty() ) {

      Log::instance()->warn( "Model file parameter will be ignored (using XML request instead)\n" );
    }
    if ( ! points_file.empty() ) {

      Log::instance()->warn( "Points file parameter will be ignored (using XML request instead)\n");
    }
  }

  // Real work

  try {

    // Load algorithms and instantiate controller class
    AlgorithmFactory::searchDefaultDirs();

    // IMPORTANT: data is not deserialized through Sampler objects, which would be much simpler.
    //            The reason is that some of the input points may be masked out so they 
    //            would be discarded when the sampler cross references occurrences and environment. 
    //            The number of evaluations MUST match the number of input points.
    AlgorithmPtr alg;
    OccurrencesPtr presences;
    OccurrencesPtr absences;
    EnvironmentPtr env;

    if ( ! request_file.empty() ) {

      // Loading input from XML request

      Log::instance()->debug( "Loading input from XML\n" );

      ConfigurationPtr input = Configuration::readXml( request_file.c_str() );

      alg = AlgorithmFactory::newAlgorithm( input->getSubsection( "Algorithm" ) );

      // Load environment and occurrences separately
      ConfigurationPtr sampler_config = input->getSubsection( "Sampler" );

      if ( ConstConfigurationPtr env_config = sampler_config->getSubsection( "Environment", false ) ) {

        env = createEnvironment();
        env->setConfiguration( env_config );
      }

      if ( ConstConfigurationPtr presences_config = sampler_config->getSubsection( "Presence", false ) ) {

        presences = new OccurrencesImpl(1.0);
        presences->setConfiguration( presences_config );
      }

      if ( ConstConfigurationPtr absences_config = sampler_config->getSubsection( "Absence", false ) ) {

        absences = new OccurrencesImpl(0.0);
        absences->setConfiguration( absences_config );
      }
    }
    else {

      // Loading input from serialized model + TAB-delimited points file

      ConfigurationPtr input = Configuration::readXml( model_file.c_str() );

      alg = AlgorithmFactory::newAlgorithm( input->getSubsection( "Algorithm" ) );

      Log::instance()->debug( "Loading training sampler to get layers, label and spatial reference\n" );

      SamplerPtr training_sampler = createSampler( input->getSubsection( "Sampler" ) );

      // IMPORTANT: environmental scenario is taken from training sampler!
      env = training_sampler->getEnvironment();

      // IMPORTANT: label and spatial reference are taken from presence points of the training sampler!
      OccurrencesPtr training_presences = training_sampler->getPresences();

      std::string label( training_presences->label() );
      std::string spatial_ref( training_presences->coordSystem() );

      Log::instance()->debug( "Loading points %s %s\n", label.c_str(), spatial_ref.c_str() );

      OccurrencesReader* oc_reader = OccurrencesFactory::instance().create( points_file.c_str(), spatial_ref.c_str() );

      presences = oc_reader->getPresences( label.c_str() );
      absences = oc_reader->getAbsences( label.c_str() );

      delete oc_reader;
    }

    if ( ! alg->done() ) {

      Log::instance()->error( "No model could be found as part of the specified algorithm. Aborting.\n");

      // If user is tracking progress
      if ( ! progress_file.empty() ) { 

        // -2 means aborted
        progressFileCallback( -2.0, &prog_data );
      }

      exit(-1);
    }

    if ( alg->needNormalization() ) {

      env->normalize( alg->getNormalizer() );
    }
    
    ConfigurationPtr output( new ConfigurationImpl("Values") );

    int num_presences = 0;
    int num_absences = 0;

    if ( presences ) {

      num_presences = presences->numOccurrences();
    }
    
    if ( absences ) {

      num_absences = absences->numOccurrences();
    }

    Log::instance()->debug( "Loaded %d presences and %d absences\n", num_presences, num_absences );
    
    // Evaluate model
    Log::instance()->debug( "Starting evaluation\n" );
    
    Scalar * vs = new (nothrow) Scalar[num_presences + num_absences];
    if ( ! vs ) {

      Log::instance()->error( "Not enough memory to allocate model values. Aborting.\n");

      // If user is tracking progress
      if ( ! progress_file.empty() ) { 

        // -2 means aborted
        progressFileCallback( -2.0, &prog_data );
      }

      exit(-1);
    }
    int cnt = 0;

    if ( presences ) {

      cnt = get_values(alg, env, presences, vs, cnt);
    }
    if ( absences ) {

      cnt = get_values(alg, env, absences, vs, cnt);
    }
    
    int precision = 5;
    
    output->addNameValue( "V", vs, num_presences + num_absences, precision );

    std::ostringstream evaluation_output;

    Configuration::writeXml( output, evaluation_output );

    std::cerr << flush;

    // Write test output to file, if requested
    if ( ! result_file.empty() ) {

      ofstream file( result_file.c_str() );
      file << evaluation_output.str();
      file.close();
    }
    else {

      // Otherwise send it to stdout
      std::cout << evaluation_output.str().c_str() << endl << flush;
    }

    // If user wants to track progress
    if ( ! progress_file.empty() ) { 

      // Indicate that the job is finished
      progressFileCallback( 1.0, &prog_data );
    }
    
    delete[] vs;
  }
  catch ( runtime_error e ) {

    // If user is tracking progress
    if ( ! progress_file.empty() ) { 

      // -2 means aborted
      progressFileCallback( -2.0, &prog_data );
    }

    printf( "om_evaluate aborted: %s\n", e.what() );
  }
}

/******************/
/*** get values ***/
int 
get_values(AlgorithmPtr alg, EnvironmentPtr env, OccurrencesPtr occs, Scalar * values, int cnt) {

  OccurrencesImpl::const_iterator oc;
  OccurrencesImpl::const_iterator end;
  Scalar val;

  if ( occs->numOccurrences() > 0 ) {

    oc = occs->begin();
    end = occs->end();
    Sample sample;

    while ( oc != end ) {
      
      // Use environmental data already provided by the point, if present
      // (in this case it will always be unnormalized, so we need to check normalization)
      if ( (*oc)->hasEnvironment() ) {

        if ( alg->needNormalization() ) {

          (*oc)->normalize( alg->getNormalizer(), env->numCategoricalLayers() );
        }
        
	sample = (*oc)->environment();
      }
      else {

        sample = env->get( (*oc)->x(), (*oc)->y() );
      }

      if ( sample.size() > 0 ) {

        //sample.dump();
        val = alg->getValue( sample );
      }
      else {
        val= -1.0;
      }

      //printf( "val=%0.5f\n", val );
      values[cnt] = val;
      ++cnt;
      ++oc;
    }
  }
  
  return cnt;
}
